### PR TITLE
ARROW-11421: [Rust][DataFusion] Support GROUP BY Date32

### PR DESCRIPTION
**Pull request by Dandandan**
After this change, TCPH 3 is running successfully:

```
Query 3 iteration 0 took 91.8 ms
Query 3 iteration 1 took 88.9 ms
Query 3 iteration 2 took 88.1 ms
Query 3 iteration 3 took 91.7 ms
Query 3 iteration 4 took 85.3 ms
Query 3 iteration 5 took 91.3 ms
Query 3 iteration 6 took 91.6 ms
Query 3 iteration 7 took 94.5 ms
Query 3 iteration 8 took 90.5 ms
Query 3 iteration 9 took 86.6 ms
Query 3 avg time: 90.03 ms
```

It is also needed by (at least) TCPH q18 but that needs more subquery support to run.

### CODE REVIEW

File: rust/datafusion/tests/sql.rs [hunk 1]
```diff
@@ -1398,6 +1401,21 @@ fn register_aggregate_simple_csv(ctx: &mut ExecutionContext) -> Result<()> {
...
+        Field::new("cnt", DataType::Int32, false),
+    ]));
+
+    ctx.register_csv(
+        "dates",
+        "tests/dates.csv",
+        CsvReadOptions::new().schema(&schema),
+    )?;
+    Ok(())
+}
```
Comment (jorgecarleitao) [inline rust/datafusion/tests/sql.rs:1417]: What do you think about placing the data in `MemTable` instead, like this:

```
    let data = RecordBatch::try_new(
        schema.clone(),
        vec![Arc::new(StringArray::from(vec![
            Some("a"),
            Some("b"),
            Some("c"),
            None,
        ]))],
    )?;
    let table = MemTable::try_new(schema, vec![vec![data]])?;
    ctx.register_table("dates", Box::new(table));
```

I find it way easier to understand the final result, in this where the numbers "6" and "9" come from.

**Review by jorgecarleitao — APPROVED** [review]
LGTM. I left a comment about readability.

File: rust/datafusion/tests/sql.rs (hunk 1 above)
Comment (Dandandan) [reply rust/datafusion/tests/sql.rs]: Thanks for the suggestion, I like it / applied it, this helps to understand the test better indeed.

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @Dandandan

**Comment by jorgecarleitao** [comment]
Thanks @Dandandan ! Could you rebase? We migrated from `Date32(_)` to `Date32` and some code here needs to be updated, unfortunately.

**Comment by Dandandan** [comment]
@jorgecarleitao already did that yesterday!

**Comment by jorgecarleitao** [comment]
uhm, the CI is complaining 

```
error[E0432]: unresolved import `arrow::datatypes::DateUnit`
  --> datafusion/tests/sql.rs:26:17
   |
26 |     datatypes::{DateUnit, TimeUnit},
   |                 ^^^^^^^^ no `DateUnit` in `datatypes`

error[E0618]: expected function, found enum variant `DataType::Date32`
    --> datafusion/tests/sql.rs:1910:28
     |
1910 |         Field::new("date", DataType::Date32(DateUnit::Day), false),
     |                            ^^^^^^^^^^^^^^^^---------------
     |                            |
     |                            call expression requires function
     |
help: `DataType::Date32` is a unit variant, you need to write it without the parenthesis
```

**Comment by Dandandan** [comment]
Ah thanks @jorgecarleitao missed that

**Comment by Dandandan** [comment]
Should be solved now @jorgecarleitao

**Review by alamb — APPROVED** [review]
This looks great @Dandandan  -- I also merged this PR into master locally and rerand the datafusion test and things are looking good. Thanks again!

**Pull request closed without merging**